Lib/xml/sax/expatreader.py

```diff
@@ -177,7 +177,7 @@ def getProperty(self, name):
         elif name == property_xml_string:
             if self._parser:
                 if hasattr(self._parser, "GetInputContext"):
-                    return self._parser.GetInputContext()
+                    return self._parser.GetInputContext().decode()
```

Comment:
`.decode()` uses the UTF-8 encoding by default. Shouldn't this be using the encoding used by the parser?

Reply:
I missed this one. Will have a new commit in a little while.

Reply:
This turns out to be much more complicated than appears on the surface. Here is a description of what happens:
* The encoding is stored in a xmlreader.InputSource instance.
* This instance is not saved in any Python class instance,
* It is passed unmodified right through to the Expat parser.
* It is permitted for the encoding to not be specified by a calling application.
* If the encoding is not specified, Expat uses the encoding specified in the XML prolog statement.
* Standard Expat only supports US-ASCII, UTF-8, UTF-16 and ISO 8859 (Latin 1)
* If the XML prolog does not specify an encoding it uses UTF-8.
* The final encoding used is not passed back to Python and Expat converts the input to UTF-8 no matter what the external coding was, losing the information about the original encoding.
* This is according to the documentation for Expat and the Python ans CPython code.

This is what will happen in various situations:
* If no encoding is specified anywhere and the document is in ISO-8859 or UTF-16, Expat will usually encounter indigestion early and will tell us. It is possible that short pathological documents might slip through.
* If no encoding is specified anywhere and the document is in US-ASCII it will be accepted correctly by Expat. If the document is in UTF-8 it will obviously be accepted.
* If the encoding is contained in the prolog of the XML document the situation is the same since Expat will reject any unsupported encodings and if the document is mis-coded it will fail in Expat as described above.
* I have not yet found where the property string which is actually the raw source line gets converted back to bytes within Python or Expat or whether Python is able to get a copy of the line before it is converted to UTF-8 by Expat for its' internal processing. Do you think I should continue researching this?.

My recommendation is to leave our code as it is as we have no way to determine what the original encoding was. Expat will handle most of the error situations itself and this feature seems to be little used by Python users based on the lack of any followup in the last 9 years to this issue.

What do you think?

Reply:
If we're going to fix this, we should fix it properly.

If expat converts the data internally to UTF-8, then the existing code here should just work, and all that will be needed is to add a comment explaining this.

If expat only supports the four encodings mentioned, then IMO we don't need to address what happens with other encodings.

I suggest adding a tests with non-ASCII characters in the XML, checking that the "XML string" property is exactly as expected, i.e. expat properly decoded it from its original encoding and then encoded into UTF-8. Do this for all encodings supported by expat (though for ASCII include only ASCII characters). This will ensure that everything works as expected.